### DISCUSSION TITLE
Getting image mean with IMagick

### DISCUSSION THREAD

**Discussion by MarliesMaalderink** [ImageMagick/ImageMagick#3714] [Help] [answered]
Hi everyone,

Is there anyone here who knows about Imagick?

I am running into a problem with IMagick (PHP), to which I am not really used, I prefer using IM on the command line, but in this case I am supposed to use the PHP library and not the command line.

I am rewriting some Java code to PHP. In it is a function in which I look up the general brightness of the image. Based on the outcome, the image is placed in one of 5 groups, and the group it is in is used to determine what should happen to the image next. 

I want the mean value of an image, the one that appears under 'Image statistics: Overall: mean:'  when I do a identify -verbose on an image. 
For a particular test image it is: 118.932 (0.466399)

But now, unfortunately, I am forced to use IMagick and I can't find it. 

If I use getImageChannelMean() with any channel or combination of channels I get values that are completely different from the values image magick returns, and I can't seem to get an overall value at all.  

If I choose to read the red channel for example, Imagick returns:
array:2 [
  "mean" => 36963.456856296
  "standardDeviation" => 6794.0036382948
]

While, for the same image, Image Magick returns: 
Red:
      min: 0  (0)
      max: 255 (1)
      mean: 151.343 (0.593503)
      standard deviation: 75.0879 (0.294462)
      kurtosis: -1.21145
      skewness: -0.257549
      entropy: 0.963865

I do not see any way in which 36963.456856296 could resemble 151.343 (0.593503)

Then Imagick also has this function: getImageChannelStatistics()

Which returns this for the same image:
array:6 [
  0 => array:5 [
    "mean" => 0.0
    "minima" => 1.7976931348623E+308
    "maxima" => -1.7976931348623E+308
    "standardDeviation" => 0.0
    "depth" => 1
  ]
  1 => array:5 [
    "mean" => 46536.614118222
    "minima" => 210.0
    "maxima" => 65535.0
    "standardDeviation" => 9048.4733238209
    "depth" => 16
  ]
  2 => array:5 [
    "mean" => 34332.323101333
    "minima" => 265.0
    "maxima" => 65474.0
    "standardDeviation" => 6089.1737130832
    "depth" => 16
  ]
  4 => array:5 [
    "mean" => 30021.433349333
    "minima" => 0.0
    "maxima" => 65531.0
    "standardDeviation" => 5244.3638779802
    "depth" => 16
  ]
  8 => array:5 [
    "mean" => 0.0
    "minima" => 1.7976931348623E+308
    "maxima" => -1.7976931348623E+308
    "standardDeviation" => 0.0
    "depth" => 1
  ]
  32 => array:5 [
    "mean" => 0.0
    "minima" => 1.7976931348623E+308
    "maxima" => -1.7976931348623E+308
    "standardDeviation" => 0.0
    "depth" => 1
  ]
]

Which seems even more illogical to me.

Anyway, I could recalculate all values I use to determine what should happen to the image based on the general brightness, but I rather not, because they are perfect the way they are and it would be complicated I guess, since I am not sure what IMagick is returning at all.

Is there anyone who can shed some light on this? How can I get the values I need from Imagick?

**Comment by snibgo** [comment]
I don't use IMagick. That system is not maintained by the ImageMagick developers.

> I do not see any way in which 36963.456856296 could resemble 151.343 (0.593503)

0.593503 is on a nominal scale of 0 to 1.

151.343 is on a nominal scale of 0 to 255. Note that 151.343 / 255 = 0.5935019607843137.

I suspect that 36963.456856296 is on a nominal scale of 0 to 65535. However, 36963.456856296 / 65535 = 0.5640261975478141, which isn't quite the same. Are you sure you are comparing like with like?

I don't know what all those "array:5" numbers mean.

**Reply by MarliesMaalderink** [reply]
Hi Snibgo,

Thank you for answering.

I double checked the numbers and did a test with a file that was pure red (or, according to the numbers, almost).

Image Magick says:
    Red:
      min: 254  (0.996078)
      max: 254 (0.996078)
      mean: 254 (0.996078)
      standard deviation: 0 (0)
      kurtosis: -1.48275e+54
      skewness: 7.5e+36
      entropy: 0

IMagick for the red channel:
  "mean" => 21759.333333333
  "standardDeviation" => NAN
  
Which seems to be not even close to 65535.
However, the green and blue channel return 0 as mean. 

Then I tried a pure magenta image (R 255 G 0 B 255) and it returned 43690.0 for the red channel, which is twice as much as a pure red one. 

So I figured I was on to something here, and I tried a pure white image, which returned 65535.0 - exactly the number you pointed out, is how far a nominal scale goes. So I guess that's what they use and I would be able to find a way to recalculate those numbers, though 

Still doesn't make any sense why a pure red image (ok, almost pure red) returns only a third of the expected number on the red channel mean. According to this graph it should have been 65535 for the red channel...
https://www.provue.com/panoramax/help/formulas_color.html

**Comment by snibgo** [comment, marked as answer]
254 (on a scale of 0 to 255) is 65278 on a scale of 0 to 65535, and 65278/3 = 21759.3333, so that number seems to be the average of all the channels, not just the red channel.

I suppose IMagick is using a scale of 0 to 65535 because your ImageMagick is Q16. If you had Q8 or Q32, I expect the maximum would be different.

See IMagick documentation https://www.php.net/manual/en/imagick.getimagechannelmean.php and https://www.php.net/manual/en/imagick.constants.php#imagick.constants.channel.

I assume you are calling...
```
getImageChannelMean(CHANNEL_RED)
```
Is that assumption correct? If not, what parameter do you use?

**Reply by MarliesMaalderink** [reply]
That is indeed what I was calling, but the results are unexpected. It seems that in some cases it does return the right value for the chosen channel (I did several tests with the colors from the table I posted in my last post), and in other cases, like with pure red, it does not. 

However, I do need the average of all channels anyway and as long as I call:

`$imagick->getImageChannelMean(Imagick::COLOR_RED | Imagick::COLOR_GREEN | Imagick::COLOR_BLUE);`

It seems to deliver the expected results. So I can work with that and make a recalculation from the old values. 

The Image Magick on my system is Q16 indeed. 

Thanks for helping me out again! (I'm not really fond of IMagick but I had no choice for this project - they seem to want to make everything just slightly different, which frustrates me, because I'm used to using Image Magick...)
